### PR TITLE
Init pmon from resource manager and read resourceEvents from kafka

### PR DESCRIPTION
**Pull request by GeorgeJahad**
# Resolves
https://jira.rax.io/browse/SALUS-135
https://jira.rax.io/browse/SALUS-117

# What
Replaces the expected keys read/watched in etcd, with the resources recieved from the resourceManager, and the kafka resourceEvent topic

## How to test
Run the tests
(Note that the tests use the embedded kafka broker in spring.)

# TODO
need to get the correct resource manager url, and figure out the hasEnvoy parameter on the  getResource() rest api call.

### CODE REVIEW

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -17,3 +24,16 @@ spring:
...
+spring:
+  profiles: test
+
+presence-monitor:
+  kafka:
+    bootstrap-servers: ${spring.embedded.kafka.brokers}
+  exportPeriod:
+    1
+  resourceManagerUrl:
+    http://localhost/getResources
```
Comment (itzg) [inline src/main/resources/application.yml]: Is this just a stub/default value or an actual location that is served by something?

File: src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java [hunk 2]
```diff
@@ -56,11 +85,10 @@ MeterRegistry getMeterRegistry() {
 
     private ObjectMapper objectMapper = new ObjectMapper();
 
-    @Autowired
+    @MockBean
     MetricExporter metricExporter;
 
-    @MockBean
-    MetricRouter metricRouter;
+    private MetricRouter metricRouter;
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java]: You could annotate that with a regular Mockito `@Mock` and avoid having to call `Mockito.mock(MetricRouter.class)` in each unit test.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        for (Map.Entry<String, PartitionSlice> e : partitionTable.entrySet()) {
+            PartitionSlice slice = e.getValue();
+            if ((record.key().compareTo(slice.getRangeMin()) >= 0) &&
+                    (record.key().compareTo(slice.getRangeMax()) <= 0)) {
+                updateSlice.accept(slice, record);
+            }
+        }
+    }
+
+    BiConsumer<PartitionSlice, ConsumerRecord<String, ResourceEvent>> updateSlice = (slice, record) -> {
```
[the remark this answers is not in the document]
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java]: @itzg I created the biConsumer so the test could wrap the biConsumer in another one that generates a semaphore.  (I don't like using sleep's in my test code.)  I can't do that with a regular method, (I don't think.)  Have I designed the test incorrectly?  How would you do it?

[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java]: Agreed, sleep's are very good to avoid. I would use Mockito's spy (and delegating) features by spying on the ResourceListener instance

https://static.javadoc.io/org.mockito/mockito-core/2.23.4/org/mockito/Mockito.html#27

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        // Prevent resources from being updated simultaneously
+        synchronized (this) {
+            String key = record.key();
+            ResourceEvent resourceEvent = record.value();
+            ResourceInfo rinfo = PresenceMonitorProcessor.convert(resourceEvent.getResource());
+            if (!resourceEvent.getOperation().equalsIgnoreCase("delete")) {
```
[the remark this answers is not in the document]
Comment (jjbuchan) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java]: https://github.com/racker/salus-telemetry-model/pull/22/files#diff-d8910738c5ff35ce0a74bc6d9d5ac8f1

File: src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java (hunk 2 above)
Comment (GeorgeJahad) [reply src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java]: @itzg But then the same mock gets reused in each test, causing the test to fail, because i need a fresh one in each test.  Am I misunderstanding something?

**Comment by GeorgeJahad** [comment]
@itzg I believe I addressed all your issues with the following exceptions:

I couldn't get spring.application.name to work and found no reference to it being the default group id in the docs, so I just used this instead:
spring.kafka.consumer.group-id: presence-monitor-${random.value}


Also, I still need to work out the hashing with James.

The diffs are here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/63d035918a3f5d63088bb47018d2229a4184bfc8...4830320980c83b0faafc3be245069f1c1cc43f51

**Comment by GeorgeJahad** [comment]
I had to add more changes based on my integration tests with james.

Things seem to be working now.  Basically I had to add the hashing, and restructure the resourcelistener to be a bean, because i stupidly failed to remember the kafka auto config only works with beans.  Anyway if you just want to see those diffs, they are here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/4830320...c47106a

**Comment by itzg** [comment]
Aha, I was confusing `spring.application.name` with a Kafka Streams config thing:
> Enabling Kafka Streams means that the application id and bootstrap servers must be set. The former can be configured using spring.kafka.streams.application-id, defaulting to spring.application.name if not set.
from [here](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-messaging.html#boot-features-kafka-streams)

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import com.rackspace.salus.telemetry.presence_monitor.services.ResourceListener;
+import com.rackspace.salus.telemetry.presence_monitor.types.PartitionSlice;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+
+@Configuration
+public class ResourceListenerBean {
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: Nit: I prefer `@Configuration` classes to end with "Config"

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    private static ConcurrentHashMap<String, PartitionSlice> partitionTable = new ConcurrentHashMap<>();
+
+    @Bean
+    public ConcurrentHashMap<String, PartitionSlice> getPartitionTable() {
+        return partitionTable;
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: `@Bean` methods are basically factory methods, so the implicit contract is that you "new" the bean inside the method. If you're wanting to ensure there's only ever one (guessing from the `static` declaration), rest assured that by default Spring bean/components are singletons and Spring will guarantee that your factory method is called only once.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Bean("resourceListener")
+    public ResourceListener getResourceListener() {
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: The convention is actually to name a `@Bean` method in non-getter former, like `resourceListener` in this case. Cool thing in this case is that you you then don't need the name specified in `@Bean`.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java [hunk 8]
```diff
@@ -84,16 +86,17 @@
                              EnvoyResourceManagement envoyResourceManagement,
                              ThreadPoolTaskScheduler taskScheduler, MetricExporter metricExporter,
                              MeterRegistry meterRegistry, KeyHashing hashing,
-                             PresenceMonitorProperties props, RestTemplateBuilder restTemplateBuilder) {
+                             PresenceMonitorProperties props, RestTemplateBuilder restTemplateBuilder,
+                             @Qualifier("resourceListener") ResourceListener resourceListener, ConcurrentHashMap<String, PartitionSlice> partitionTable) {
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java]: Hmm, why did you need the `@Qualifier`? Was it complaining about ambiguous instances?

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java [hunk 9]
```diff
@@ -43,11 +47,13 @@ public void resourceListener(ConsumerRecord<String, ResourceEvent> record) {
         boolean keyFound = false;
         for (Map.Entry<String, PartitionSlice> e : partitionTable.entrySet()) {
             PartitionSlice slice = e.getValue();
-            if ((record.key().compareTo(slice.getRangeMin()) >= 0) &&
-                    (record.key().compareTo(slice.getRangeMax()) <= 0)) {
+            ResourceInfo rinfo = PresenceMonitorProcessor.convert(record.value().getResource());
+            String hash = PresenceMonitorProcessor.genExpectedId(rinfo);
+            if ((hash.compareTo(slice.getRangeMin()) >= 0) &&
```
Comment (itzg) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java]: Sorry, I forgot to mention this yesterday, but it would be handy to have a method on `PartitionSlice` like a `containsHash` that will do this range conditional. Then you can simply these statements.

File: src/main/resources/application.yml [hunk 10]
```diff
@@ -7,11 +7,11 @@ presence-monitor:
     EVENT:
       telemetry.events.json
     RESOURCE:
-      telemetry.resources.json
+      resources
```
Comment (itzg) [inline src/main/resources/application.yml]: Actually, I prefer the `telemetry.` prefixing type structure of topic name. I keep forgetting to mention to @jjbuchan about the semi-established topic naming convention.

File: src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java [hunk 11]
```diff
@@ -127,11 +126,19 @@ MeterRegistry getMeterRegistry() {
...
+    @Autowired
+    @Qualifier("resourceListener")
+    ResourceListener resourceListener;
+
     @Mock
     ClientHttpResponse response;
 
     @Before
     public void setUp() throws Exception {
+        partitionTable.forEachKey(1, k -> partitionTable.remove(k));
```
Comment (itzg) [inline src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java]: You could also use `@DirtiesContext` on the test class or test methods to indicate that a fresh set of components is needed for each test.

If staying with this approach, which is totally fine, I'd recommend just use the [clear method](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#clear--)

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java (hunk 6 above)
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: @itzg The problem is I want the PresenceMonitorProcessor class to share the partitionTable with the resourceListener

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java (hunk 8 above)
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java]: yes, the test code also creates a bean so that spring can autoconfig it with embedded kafka broker.

File: src/main/resources/application.yml (hunk 10 above)
Comment (GeorgeJahad) [reply src/main/resources/application.yml]: yes, i just changed it for the integration test with james code.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java (hunk 6 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: They would both get the same, singleton instance. Spring will only call the `@Bean` factory once and then provide the returned instance to all uses of it.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java (hunk 7 above)
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: When I remove it I get this error:
Unsatisfied dependency expressed through constructor parameter 9; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.rackspace.salus.telemetry.presence_monitor.services.ResourceListener' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Qualifier(value=resourceListener)}

Comment (itzg) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: You'll need to make both changes I suggested there to end up with

```
    @Bean
    public ResourceListener resourceListener() {
        return new ResourceListener(partitionTable);
    }
```

File: src/main/resources/application.yml (hunk 10 above)
Comment (jjbuchan) [reply src/main/resources/application.yml]: locally i've changed to telemetry.resources.json and telemetry.attaches.json

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java (hunk 7 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java]: FYI the bean naming behavior of `@Bean` is described deep in the docs here

https://docs.spring.io/spring/docs/5.1.4.RELEASE/spring-framework-reference/core.html#beans-java-bean-annotation

**Comment by GeorgeJahad** [comment]
ok, i think I got everything:

diffs here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/c47106a0b294d467f725f828e623f0f2adeada18...george/addResourceManager3

**Comment by GeorgeJahad** [comment]
@itzg Thank you for taking the time to teach me all of this stuff.  There are so many subtle spring details that I would have missed if not for you.

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java [hunk 12]
```diff
@@ -70,6 +112,51 @@ private String getExpectedId(KeyValue kv) {
...
+                        }
+                }
+                return response;
+            });
+            return resources;
+        }
+    }
+
+    static ResourceInfo convert(Resource resource) {
+        ResourceInfo ri = new ResourceInfo();
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java]: Doesn't need to change, but I did like the additional features this provided - https://github.com/racker/salus-telemetry-resource-management/pull/1/files#diff-d034cbc23c60da4b9dc850475e9f9768R218

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java [hunk 13]
```diff
@@ -95,23 +183,24 @@ public void start(String id, String content) {
...
-            PartitionSlice.ExpectedEntry expectedEntry = new PartitionSlice.ExpectedEntry();
-            try {
-                resourceInfo = objectMapper.readValue(kv.getValue().getBytes(), ResourceInfo.class);
-            } catch (IOException e) {
-                log.warn("Failed to parse ResourceInfo", e);
+        List<Resource> resources = getResources();
+
+        log.debug("Found {} expected envoys", resources.size());
+        resources.forEach(resource -> {
+            if (!resource.getPresenceMonitoringEnabled())
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java]: The endpoint your hitting only returns resources with this enabled, so this check should be irrelevant

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/types/KafkaMessageType.java [hunk 14]
```diff
@@ -21,5 +21,6 @@
 public enum KafkaMessageType {
     LOG,
     METRIC,
-    EVENT
+    EVENT,
+    RESOURCE
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/telemetry/presence_monitor/types/KafkaMessageType.java]: There is a shared `KafkaMessageType` in the models repo you can use.  Under `com.rackspace.salus.telemetry.messaging`

File: src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java [hunk 15]
```diff
@@ -1,49 +1,79 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
 
 @RunWith(SpringRunner.class)
+@EnableAutoConfiguration
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE)
-
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
Comment (jjbuchan) [inline src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java]: Interesting

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java (hunk 13 above)
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java]: will do

File: src/main/java/com/rackspace/salus/telemetry/presence_monitor/types/KafkaMessageType.java (hunk 14 above)
Comment (GeorgeJahad) [reply src/main/java/com/rackspace/salus/telemetry/presence_monitor/types/KafkaMessageType.java]: will do
